src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -1890,3 +1969,91 @@ fn is_primitive(path_str: &str, ns: Namespace) -> Option<(&'static str, Res)> {
 fn primitive_impl(cx: &DocContext<'_>, path_str: &str) -> Option<&'static SmallVec<[DefId; 4]>> {
     Some(PrimitiveType::from_symbol(Symbol::intern(path_str))?.impls(cx.tcx))
 }
+
+fn strip_generics_from_path(path_str: &str) -> Result<String, ResolutionFailure<'static>> {
+    let mut stripped_segments = vec![];
+
+    let mut path = path_str.chars().peekable();
+
+    let mut segment = Vec::new();
+
+    while let Some(chr) = path.next() {
+        match chr {
+            ':' => {
+                if path.next_if_eq(&':').is_some() {
+                    let stripped_segment =
+                        strip_generics_from_path_segment(mem::take(&mut segment))?;
+                    if !stripped_segment.is_empty() {
+                        stripped_segments.push(stripped_segment);
+                    }
+                } else {
+                    return Err(ResolutionFailure::MalformedGenerics(
+                        MalformedGenericsVariant::InvalidPathSeparator,
+                    ));
+                }
+            }
+            '<' => {
+                segment.push(chr);
+                while path.peek().map_or(false, |c| *c != '>') {
```

Comment:
This will be wrong for `Box<Vec<T>>` I think. It will see the first `<` and stop at the first `>`, so the last `>` will be left over and give an error. Try keeping an explicit count of the number of `<` open.

Also you can use `next_if` here too:
```suggestion
                while let Some(ch) = path.next_if(|c| *c != '>') {
```

Reply:
Another thing: this won't warn on `"<T as S>"`. Maybe consider unifying this with `strip_generics_from_path_segment` somehow?

Reply:
It does warn:

```text
error: unresolved link to `<T as S>`
  --> $DIR/intra-link-malformed-generics.rs:21:6
   |
LL | //! [<T as S>]
   |      ^^^^^^^^ fully-qualified syntax is unsupported
   |
   = note: see https://github.com/rust-lang/rust/issues/74563 for more information
```

Reply:
Hmm, I'm not understand the code then ... let me re-read it.

Reply:
> This will be wrong for `Box<Vec<T>>` I think.

It does work. Here's what I think will happen:

1. It eats `Box` and stores that in `segment`
2. It sees `<`, continues adding to `segment` until it reaches the `>` after `T`, and then stops before it consumes the `>`
3. It sees `>`, consumes that as a regular character, and adds it to `segment`

Reply:
The integration test covers a similar case :)

https://github.com/rust-lang/rust/blob/e35e340603512b0b8ca1db01f6239ec7d6ddf123/src/test/rustdoc/intra-doc-link-generic-params.rs#L5

Reply:
Oh I see - it works because it calls `strip_generics` _twice_ - once if it sees `::` and once after it runs into the end of the string. Hmm - try this test case:

```rust
 //! Here's a link to [`Box<Vec<T>>::new`]. 
```

Reply:
Added the test case.

Reply:
`new` is a bad example because it's a semantic error, but it would be valid for associated items.

Reply:
I made it semantically valid :)